onnxruntime/test/util/default_providers.cc

```diff
@@ -16,7 +16,9 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_CUDA(
     OrtCudnnConvAlgoSearch cudnn_conv_algo = OrtCudnnConvAlgoSearch::EXHAUSTIVE,
     size_t cuda_mem_limit = std::numeric_limits<size_t>::max(),
     ArenaExtendStrategy arena_extend_strategy = ArenaExtendStrategy::kNextPowerOfTwo,
-    bool do_copy_in_default_stream = true);
+    bool do_copy_in_default_stream = true,
+    void* external_alloc = nullptr,
```

Comment:
we should group these two fields into a Struct, e.g. ExternalAllocInfo. 
I think we will probably need to enrich this struct latter.

Reply:
for example, imagine if you were to include empty_cache() API in the allocator. Having a struct means you don't need to change all these function interface in these many files.

Reply:
I agree with the general direction but these function pointers are already part of CUDAExecutionProviderInfo which is a struct. Unfortantely, CreateExecutionProviderFactory_CUDA is not exposed via header file, hence all of its callers have to define it as an external function. Given at some point we will have a way to register external allocators I'm not sure it makes sense to wrap these function pointers in another struct because that would mean we will need to redefine these struct is all the files that call  CreateExecutionProviderFactory_CUDA

---
In reply to: [568314234](https://github.com/microsoft/onnxruntime/pull/6529#discussion_r568314234) [](ancestors = 568314234)